### PR TITLE
BUG: Escape single quote in DICOM database SQL statement

### PR DESCRIPTION
**Pull request by cpinter**
Re https://issues.slicer.org/view.php?id=4709

### CODE REVIEW

**Comment by lassoan** [comment]
Looks good, but why don't we do it with all the other fields? For example, physician's name can contain single-quote, too.

**Comment by cpinter** [comment]
It occurred to me but didn't seem likely even for the physician name. If you give me a list of fields you want me to escape the same way I'll add those.

**Comment by lassoan** [comment]
It is very common in person names (O'Leary...) and can easily occur in any description fields. Are there any other restricted characters? Does \ have to be encoded as \\?

**Comment by pieper** [comment]
I don't think this is the source of the issue - the data was imported correctly in the first place, but the error happened when it converted to the new schema.  Can you point to the commits involved in the database update?

**Comment by cpinter** [comment]
Sorry I had only a little time in the evening and didn't pay enough attention. I'll track this down now.

**Comment by cpinter** [comment]
Actually it would help a lot if I had that data and could debug into it. Without it it's just guesswork.

**Comment by cpinter** [comment]
Anyway, I did some guesswork and pushed a new change :) It escapes the single quote character in every field that is being processed during update.

**Comment by pieper** [comment]
Thanks for digging into this @cpinter 

I still don't understand why this became an issue for converting when the data imported correctly before.  Do you have any ideas on that?  Without that it makes it hard to know if the fix is correct.

I can't share the exact dataset (PHI) and I haven't had time to check, but I can do more debugging/testing at some point as needed.  For now I'd be happy just understanding what changed - can you point to the commits that handle the database update?

**Comment by cpinter** [comment]
I also don't understand how the initial insertion can be fine without having to sanitize, but then fails later.

This is the original commit
https://github.com/commontk/CTK/commit/9c2af28e84da1abb986036317d75009d4c149923
And a few follow-up commits - basically whatever I committed to CTK lately. This is quite a big change, and I'm not sure how reading it would help. If you have specific questions I'm happy to answer though.

**Comment by pieper** [comment]
Yes, I was looking through and it seems that this squashed commit has everything lumped together and I don't see what would have led to this issue.

Specifically though if you look at the error message it's in an `UPDATE` to the `Studies` table - do you know where that would happening?

```
 Bad SQL: UPDATE Studies SET DisplayedNumberOfSeries='3', InstitutionName='BRIGHAM & WOMEN'S HOSPITAL', PatientsUID=4, ReferringPhysician='x^x^^^', StudyDate='2019xxxx', StudyDescription='MRI BRAIN FUNCTIONAL', StudyInstanceUID='1.2.xx' WHERE StudyInstanceUID='1.2.x'
```

**Comment by cpinter** [comment]
This PR should fix the issue.

The updated values are set here: https://github.com/commontk/CTK/blob/master/Libs/DICOM/Core/ctkDICOMDatabase.cpp#L929
And then the update command is called here:
https://github.com/commontk/CTK/blob/master/Libs/DICOM/Core/ctkDICOMDatabase.cpp#L1066

**Comment by cpinter** [comment]
Maybe this PR is not enough after all. We need to make sure the displayed field generator does not override the original value with some problematic value. It's here:
https://github.com/commontk/CTK/blob/master/Libs/DICOM/Core/ctkDICOMDisplayedFieldGenerator.cpp#L116

**Comment by pieper** [comment]
Oh well yes, this is where the problem is: https://github.com/commontk/CTK/blob/master/Libs/DICOM/Core/ctkDICOMDatabase.cpp#L1066

This uses QString formatting to build the SQL query instead of the usual `prepare` and `bind` that is used elsewhere.  It should be changed to match the rest of the QSqlQuery usage and the problem should go away.

**Comment by pieper** [comment]
Also line 1144 and anywhere else where sql statements are created.

**Comment by cpinter** [comment]
OK. I'll change the queries to use prepare everywhere in this class.

**Comment by cpinter** [comment]
Line 1144 seems to be just a closing brace...

**Comment by pieper** [comment]
> Line 1144 seems to be just a closing brace...

Sorry, meant 1104, but you have the idea. 😄 

I don't have the exact commit handy but I ran into this exact issue with the code when it was first written back seven or eight years ago when first testing it with data from `Brigham and Women's Hospital` and that's why we changed to using the `prepare` and `bind` calls.

**Comment by cpinter** [comment]
One question, maybe you know this by heart. The prepare/bind way just replaces strings, or is there something "more intelligent" going on there? Because the displayStudiesFieldUpdateList and its similar counterparts are longer strings involving multiple fields, and if it's just string replacement then it's fine, but if it somehow forces to bind values for fields, then it probably won't work like this.

**Comment by cpinter** [comment]
I updated the PR. It is not working yet (crashes), but it will be something like this. I have to go now though, sop I'll fix this tomorrow.

**Comment by cpinter** [comment]
This should do it. @pieper can you please test with your dataset?

**Comment by pieper** [comment]
Hmm, sorry, no, still not working.  But it's closer.  Now the error is:

```
SQL failed
 Bad SQL: UPDATE Studies SET DisplayedNumberOfSeries = ? , InstitutionName = ? , PatientsUID = ? , ReferringPhysician = ? , StudyDate = ? , StudyDescription = ? , StudyInstanceUID = ? StudyInstanceUID UID = ? ;
Error text:  Parameter count mismatch
```

File: Libs/DICOM/Core/ctkDICOMDatabase.cpp [hunk 1]
```diff
@@ -1038,38 +1044,47 @@ bool ctkDICOMDatabasePrivate::applyDisplayedFieldsChanges( QMap<QString, QMap<QS
...
       // Trim the separators from the end
-      displayStudiesFieldUpdateList = displayStudiesFieldUpdateList.left(displayStudiesFieldUpdateList.size() - 2);
+      displayStudiesFieldUpdateString = displayStudiesFieldUpdateString.left(displayStudiesFieldUpdateString.size() - 3);
 
-      QSqlQuery updateDisplayStudyStatement(this->Database);
-      QString updateDisplayStudyStatementString = 
-        QString("UPDATE Studies SET %1 WHERE StudyInstanceUID='%2';").arg(displayStudiesFieldUpdateList).arg(currentStudy["StudyInstanceUID"]);
-      this->loggedExec(updateDisplayStudyStatement, updateDisplayStudyStatementString);
+      QSqlQuery updateDisplayStudyStatement(this->Database);
+      updateDisplayStudyStatement.prepare( QString("UPDATE Studies SET %1 StudyInstanceUID UID = ? ;").arg(displayStudiesFieldUpdateString) );
```
Comment (pieper) [inline Libs/DICOM/Core/ctkDICOMDatabase.cpp]: Doesn't this sql statement need a WHERE clause?

**Comment by pieper** [comment]
I took a look but didn't see anything obvious here about the parameter counts, but maybe it's related to the comment I added in the code about a missing WHERE clause?  We can chat again about it if you want to walk through the logic together.

**Comment by cpinter** [comment]
Interesting. I tried a simple import then update and it worked for me. Based on the error message, a WHERE seems to be missing. Not sure what's going on. Is it OK if I check it out on Tuesday?

**Comment by pieper** [comment]
Oh yes, Tuesday sounds great!

**Comment by cpinter** [comment]
I changed the statement. I think I accidentally pasted the StudyInstanceUID on the WHERE instead the UID... Not sure how it worked when I tried it. I hope it's going to be good now.

**Comment by pieper** [comment]
Tested and it looks good to me - thanks @cpinter! 👍

**Comment by pieper** [comment]
Okay, I updated Slicer to use this, so we should be good to go.  I'll try to check in the nightly build.

**Comment by cpinter** [comment]
Thank you very much! Sorry for the problems!

**Comment by jcfr** [comment]
:tada:

**Pull request merged**